spec/puppet/resource_api/property_spec.rb

```diff
@@ -0,0 +1,126 @@
+require 'spec_helper'
+
+RSpec.describe Puppet::ResourceApi::Property do
+  subject(:property) do
+    described_class.new(type_name, data_type, attribute_name, resource_hash)
+  end
+
+  let(:type_name) { 'test_name' }
+  let(:attribute_name) { 'some_property' }
+  let(:data_type) { Puppet::Pops::Types::PStringType.new(nil) }
+  let(:resource_hash) { { resource: {} } }
+
+  describe '#new(type_name, data_type, attribute_name, resource_hash)' do
+    it { expect { described_class.new(nil) }.to raise_error ArgumentError, %r{wrong number of arguments} }
+    it { expect { described_class.new(type_name, data_type, attribute_name, resource_hash) }.not_to raise_error }
+
+    context 'when name is :ensure' do
+      let(:attribute_name) { :ensure }
+
+      it 'the #insync? method is avaliable' do
+        expect(property.methods.include?(:insync?)).to eq true
+      end
+    end
+  end
+
+  describe '#should=(value)' do
```

Comment:
I think the `should`/`should=`/`value`/`rs_value` tests can be simplified by folding them all into a unified structure:

```
describe 'value munging and storage' do
  before(:each) do
    allow(Puppet::ResourceApi::DataTypeHandling).to receive(:mungify)
      .with(data_type, value, 'test_name.some_property', false)
      .and_return(munged_value)

    property.should = value
  end

  context 'when handling strings' do
    let(:value) { 'value' }
    let(:munged_value) { 'munged value' }

    it { expect(property.should).to eq 'munged value' }
    it { expect(property.rs_value).to eq 'munged value' }
    it { expect(property.value).to eq 'munged value' }
  end

  context 'when handling boolean true' do
    let(:value) { true }
    let(:munged_value) { true }
    let(:data_type) { Puppet::Pops::Types::PBooleanType.new }

    it { expect(property.should).to eq :true }
    it { expect(property.rs_value).to eq true }
    it { expect(property.value).to eq :true }
  end
end
```

That gets rid of poking into the class' private parts with `instance_variable_set`, and clearly defines the expectations around how the different interfaces (should/value/rs_value) return the stored value.

Reply:
This proposed tests look much cleaner and nicer. 😄

Reply:
There is also this `:ensure` handling that needs to be tested: https://github.com/puppetlabs/puppet-resource_api/blob/9dd87a93c8e0926a90745058df5005d311825426/lib/puppet/resource_api.rb#L232